### PR TITLE
[basic.stc.dynamic.general] Fix definition/explanation of dynamic storage duration

### PR DESCRIPTION
**Pull request by eisenwave**
As part of https://github.com/cplusplus/draft/issues/7261.

This PR fixes numerous issues with [[basic.stc.dynamic.general] p1](https://eel.is/c++draft/basic.stc.dynamic.general#1):

1. "dynamically" is not defined in the standard, and the paragraph does not define it (judging by formatting), so this term is meaningless.

2. The paragraph specifically states "during program execution", but C++20 allows the user to create dynamic storage duration objects during *program translation*, i.e. within constant evaluation.

3. The explanation given here is not exhaustive. Normative wording in [[basic.stc.general] p2](https://eel.is/c++draft/basic.stc.general#2.sentence-2) already states "The dynamic storage duration is associated with objects created by a *new-expression* ([expr.new]) or with implicitly created objects ([intro.object]).". Notably, implicitly created objects are not mentioned in [basic.stc.dynamic.general].

It is worth noting that the new definition explicitly gives dynamic storage duration to *objects*, which is inconsistent with the other definitions. For example, [[basic.stc.static]](https://eel.is/c++draft/basic.stc.static) says:

> All variables which [...] have *static storage duration*.

@jensmaurer has stated in https://github.com/cplusplus/draft/issues/6523#issuecomment-1694706321 that we have long-standing problems with the term "variable". At first sight, this may be a blocker for this edit.

However, this editorial issue does not introduce a novel problem because [[expr.new] p10](https://eel.is/c++draft/expr.new#10) already states:

> Objects created by a *new-expression* have dynamic storage duration ([basic.stc.dynamic]).

In other words, this PR makes no normative changes. It aggregates existing wording from [expr.new] and [basic.stc.general] and replaces old incorrect or incomplete wording.

### CODE REVIEW

File: source/basic.tex [hunk 1]
```diff
@@ -3989,15 +3989,15 @@
 \rSec3[basic.stc.dynamic]{Dynamic storage duration}%
 
 \rSec4[basic.stc.dynamic.general]{General}%
-\indextext{storage duration!dynamic|(}
```
Comment (jensmaurer) [inline source/basic.tex:3992]: Why is this being removed?

File: source/basic.tex [hunk 2]
```diff
@@ -3989,15 +3989,15 @@
...
 
 \pnum
-Objects can be created dynamically during program
-execution\iref{intro.execution}, using
+Object with \defnadj{dynamic}{storage duration} are created using
```
Comment (jensmaurer) [inline source/basic.tex:3994]: This seems an English grammar bug: "Object ... are"

Comment (jensmaurer) [inline source/basic.tex:3994]: Can we reverse the phrasing here?  The storage duration is a consequence of the syntax used to create the object, not the other way round: "Objects created using X and Y have dynamic storage duration."  That seems to be the way we phrase it for the other storage durations (never mind object vs. variable).

**Review by jensmaurer — CHANGES_REQUESTED** [review]

File: source/basic.tex (hunk 1 above)
Comment (Eisenwave) [reply source/basic.tex:3992]: The PR adds `\defnadj` within the paragraph below, which presumably obsoletes this.

File: source/basic.tex (hunk 2 above)
Comment (Eisenwave) [reply source/basic.tex]: Done. The only reasonable way to accommodate the note on *delete-expression*s was to pull it into a completely separate sentence though, which is why I originally refrained from doing that.

I think it's overall for the better now though; good suggestion.

File: source/basic.tex (hunk 1 above)
Comment (jensmaurer) [reply source/basic.tex:3992]: There's a similar thing with a closing paren at the end of the section (I think), which pairs with this one.  You can't just remove one and leave the other. I'd like to point out that remove those parens gets rid of a "n-m" index entry, which we might want to keep.

Comment (eisenwave) [reply source/basic.tex:4474]: I've added it back in.
